File: problems/74-search-a-2d-matrix/solution.ts
function searchA2DMatrix(matrix: number[][], target: number) {
  let firstRow = 0;
  let lastRow = matrix.length;

  while (firstRow <= lastRow) {
    let midRow = Math.floor((firstRow + lastRow) / 2);

    if (
      matrix[midRow][0] <= target &&
      matrix[midRow][matrix.length - 1] >= target
    ) {
      let left = 0;
      let right = matrix[midRow].length;

      while (left <= right) {
        let mid = Math.floor((left + right) / 2);

        if (matrix[midRow][mid] === target) return true;
        else if (matrix[midRow][mid] > target) right = mid - 1;
        else left = mid + 1;
      }

      return false;
    } else if (matrix[midRow][0] > target) lastRow = midRow - 1;
    else firstRow = midRow + 1;
  }

  return false;
}

export { searchA2DMatrix };

File: problems/74-search-a-2d-matrix/solution.go
package main

func searchA2DMatrix(matrix [][]int, target int) bool {
	var rowContainsRange = []int{}

	var first, last = 0, len(matrix) - 1

	loop: for first <= last {
		var midRow int = (first + last) / 2

		switch {
		case matrix[midRow][0] > target:
			last = midRow - 1
		case matrix[midRow][len(matrix[midRow])-1] < target:
			first = midRow + 1
		default:
			rowContainsRange = matrix[midRow]
			break loop
		}
	}

	if first > last {
		return false
	}

	first, last = 0, len(rowContainsRange)-1

	for first <= last {
		var mid int = (first + last) / 2

		switch {
		case rowContainsRange[mid] == target:
			return true
		case rowContainsRange[mid] > target:
			last = mid - 1
		default:
			first = mid + 1
		}
	}

	return false
}
